src/vs/base/common/assertDefined.ts

```diff
@@ -0,0 +1,28 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+/**
+ * Asserts that a provided `value` is `defined` - not `null` or `undefined`,
+ * throwing an error with the provided error or error message otherwise.
+ *
+ * @throws if the provided `value` is `null` or `undefined`.
+ *
+ * ## Examples
+ *
+ * ```typescript
+ * // an assert with an error message
+ * assertDefined('some value', 'String constant is not defined.')
+ *
+ * // `throws!` an assert with an error message
+ * assertDefined(null, new Error('Should throw this error.'))
+ * ```
+ */
+export function assertDefined<T>(value: T, error: string | NonNullable<Error>): asserts value is NonNullable<T> {
```

Comment:
This one is a little bit different since it is an `assert` and doesn't require variable reassignment. Let me see if I can use the existing one without need to create `let` variable just for the assertion 🤔

Reply:
I've looked at it and it does not really work in my use cases 😞 The main issue is that the existing on does not scope down type of the variable.

Real-world example:

```typescript
export type IChatRequestVariableValue = string | URI | Location | unknown | Uint8Array;

function example(data: IChatRequestVariableValue) {
    data = assertIsDefined(data);
    data // 💥 data is still `unknown` hence nothing really changes type-wise
}
```

Hence I can remove this util, but I had to add manual type checks/throw errors everywhere which will be even more verbose 🤷

Reply:
Oh ok, well maybe this can still go in `types.ts` since it seems to fit in with those helpers. Or maybe one of them can be adapted for this use case